docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,398 @@
+	GetSecrets(path string, _ ...string) (map[string]string, error)
+    // Sets the secrets lastupdated time to current time. 
+    // Needed for InsecureSecrets support
+	SecretsUpdated()
+    // Returns the secrets last updated time
+	SecretsLastUpdated() time.Time
+}
+```
+
+> *Note: The `GetDatabaseCredentials` and `GetCertificateKeyPair` APIs have been removed. These are no longer needed since insecure database credentials will no longer be stored in the `DatabaseInfo` configuration and certificate key pairs are secrets like any others. This allows these secrets to be retrieved via the `GetSecrets` API.*
+
+### Implementation
+
+#### Factory Method and Bootstrap Handler
+
+The factory method and bootstrap handler will follow that currently in the Bootstrap implementation with some tweaks. Rather than putting the two split interfaces into the DIC, it will put just the single interface instance into the DIC. See details in the [Interfaces and factory methods](#interfaces-and-factory-methods) section above under **Existing Implementations**.
+
+#### Caching of Secrets
+
+Secrets will be cached as they are currently in the Application Service implementation
+
+#### Insecure Secrets
+
+Insecure Secrets will be handled as they are currently in the Application Service implementation. `DatabaseInfo` configuration will no longer be an option for storing the insecure database credentials. They will be stored in the `InsecureSecrets` configuration only.
+
+```toml
+[Writable.InsecureSecrets]
+    [Writable.InsecureSecrets.DB]
+    	path = "redisdb"
+    	[Writable.InsecureSecrets.DB.Secrets]
+            username = ""
+            password = ""
+```
+
+##### Handling on-the-fly changes to `InsecureSecrets`
+
+All services will need to handle the special processing when `InsecureSecrets` are changed on-the-fly via Consul. Since this will now be a common configuration item in `Writable` it can be handled in `go-mod-bootstrap` along with existing log level processing. This special processing will be taken from App SDK.
+
+#### Mocks
+
+Proper mock of the `SecretProvider` interface will be created with `Mockery` to be used in unit tests. Current mock in App SDK is handle written rather then generated with `Mockery`.
+
+#### Where will `SecretProvider` reside?
+
+##### Go Services
+
+The final decision to make is where will this new `SecretProvider` abstraction reside? Originally is was assumed that it would reside in `go-mod-secrets`, which seems logical. If we were to attempt this with the implementation including the bootstrap handler, `go-mod-secrets` would have a dependency on `go-mod-bootstrap` which will likely create a circular dependency. 
+
+Refactoring the existing implementation in `go-mod-bootstrap` and have it reside there now seems to be the best choice.
+
+##### C Device Service
+
+The C Device SDK will implement the same `SecretProvider` abstraction, InsecureSercets configuration and the underling `SecretStore` client.
+
+### Consequences
+
```

Comment:
We might add one more consequence of the `go-mod-secrets` position.  Do we still need it?

Reply:
go-mod-secrets is still needed for the SecretClient that wraps the Vault requests.